### ISSUE TITLE
Avoid using PATH_MAX

### ISSUE THREAD

**Issue by egrimley** [DynamoRIO/dynamorio#2110]
PATH_MAX is used in a couple of places in `dr_frontend_unix.c`, but it is best avoided. Not all systems define it, and even if it is defined it could have an unhelpful value such as INT_MAX.

Modern implementations of realpath() allow the second argument to be NULL. If it is necessary to cope with ancient implementations a wrapper like this could be used:

```
/* Handle the (unlikely) case of us having an ancient realpath() that does
 * not accept a second argument that is NULL.
 */
static char *
realpath_wrapper(const char *path)
{
    char *result = realpath(path, NULL);
#if defined(PATH_MAX) && PATH_MAX > 0
    if (result == NULL && path != NULL && errno == EINVAL) {
        /* Try again with a fixed-length buffer. */
        char *buf = malloc(PATH_MAX);
        if (buf != NULL) {
            result = realpath(path, buf);
            if (result != NULL) {
                /* Copy result into malloc-ed memory, if possible. */
                char *tmp = malloc(strlen(result) + 1);
                if (tmp != NULL)
                    strcpy(tmp, result);
                result = tmp;
            }
            free(buf);
        }
    }
#endif
    return result;
}
```

**Comment by derekbruening** [comment]
Our code uses MAXIMUM_PATH everywhere.  These PATH_MAX instances should be MAXIMUM_PATH instead.

**Comment by egrimley** [comment]
In drfront_searchenv, you definitely shouldn't call realpath() with the second argument set to a buffer of length MAXIMUM_PATH while MAXIMUM_PATH is defined to be only 260: realpath() could overflow the buffer.

In drfront_dir_try_writable, there's no justification for using any kind of fixed-length buffer. You can allocate exactly the right amount: strlen(path) + strlen(TMP_FILE_NAME) + 2.

**Comment by derekbruening** [comment]
Hmm, I thought MAXIMUM_PATH was larger than 260 for UNIX, and in fact recently I ran into it being too small: #2116.

**Comment by egrimley** [comment]
If you want to port DynamoRIO to GNU Hurd you'll have to cope without a maximum path: https://www.gnu.org/software/hurd/hurd/porting/guidelines.html#PATH_MAX_tt_MAX_PATH_tt_MAXPATHL

However, since there are about 300 mentions of MAXIMUM_PATH in the DynamoRIO source code, probably you'll need a stronger motivation than GNU-Hurd-compatibility for rewriting DynamoRIO to dynamically allocate all filepaths...

Linux seems to enforce a limit of 4095 characters (without the null byte). If you try to sys_open a longer path you get ENAMETOOLONG, even if the file exists. I would guess that for most software coping *correctly* with all the corner-cases created by having a maximum path length is more difficult than dynamically allocating the file names and not having the corner cases.

**Comment by derekbruening** [comment]
Yes, we are aware of the 4096 Linux limit (see comments in globals_shared.h).

It is not easy to dynamically allocate in several places where we need buffers that contain file paths, otherwise we would have done it long ago.

**Comment by egrimley** [comment]
To demonstrate that it is not always possible to canonicalise a valid filepath:

```
cd /tmp
x255=$(perl -e 'print "x" x 255;')
x=0 ; while [ $x -lt 20 ] ; do mkdir $x255 ; ln -s $x255 y ; cd y ; x=$[$x+1] ; done
```

The current directory is now /tmp/y/y/y/y/y/y/y/y/y/y/y/y/y/y/y/y/y/y/y/y, and everything seems to work nicely there, but the real path would have 5124 characters.
